crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -584,11 +588,103 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
                 let output = self.object_summary_output(&summary, view_events, view_gas_used);
                 Ok(output)
             }
+            SuiSubcommand::UpgradePackage(UpgradePackageCommand {
+                package,
+                upgrade_capability,
+                gas_budget,
+            }) => {
+                let data = data.ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "Expected a module text block following 'upgrade' starting on lines {}-{}",
+                        start_line,
+                        command_lines_stop
+                    )
+                })?;
+                let (units, warnings_opt) = self.compile_source_units(data.path())?;
+                let modules = units
+                    .into_iter()
+                    .map(|unit| match unit {
+                        AnnotatedCompiledUnit::Module(annot_module) => {
+                            let (named_addr_opt, _id) = annot_module.module_id();
+                            let named_addr_opt = named_addr_opt.map(|n| n.value);
+                            let module = annot_module.named_module.module;
+                            (named_addr_opt, module)
+                        }
+                        AnnotatedCompiledUnit::Script(_) => panic!(
+                            "Expected a module text block, not a script, \
+                                    following 'publish' starting on lines {}-{}",
+                            start_line, command_lines_stop
+                        ),
+                    })
+                    .collect();
+
+                let output =
+                    self.upgrade_package(package, modules, upgrade_capability, gas_budget)?;
+                Ok(merge_output(warnings_opt, output))
+            }
```

Comment:
Inspired by `move-lang/move` for publish: https://github.com/move-language/move/pull/1017/files. Note: I punted on casing out `SyntaxChoice` for `IR` and just assume source. Do we want to support IR here?